gemini/function-calling/function_calling_service/main.py

```diff
@@ -0,0 +1,178 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import os
+
+from flask import Flask, jsonify, render_template, request
+import requests
+import vertexai
+from vertexai.generative_models import (
+    Content,
+    FunctionDeclaration,
+    GenerationConfig,
+    GenerativeModel,
+    Part,
+    Tool,
+)
+
+logger = logging.getLogger(__name__)
+
+get_location = FunctionDeclaration(
+    name="get_location",
+    description="Get latitude and longitude for a given location",
+    parameters={
+        "type": "object",
+        "properties": {
+            "poi": {"type": "string", "description": "Point of interest"},
+            "street": {"type": "string", "description": "Street name"},
+            "city": {"type": "string", "description": "City name"},
+            "county": {"type": "string", "description": "County name"},
+            "state": {"type": "string", "description": "State name"},
+            "country": {"type": "string", "description": "Country name"},
+            "postal_code": {"type": "string", "description": "Postal code"},
+        },
+    },
+)
+
+location_tool = Tool(
+    function_declarations=[get_location],
+)
+
+model = GenerativeModel(
+    "gemini-1.5-flash-001",
+    generation_config=GenerationConfig(temperature=0),
+    tools=[location_tool],
+)
+
+app = Flask(__name__)
+
+PROJECT_ID = os.environ.get("GOOGLE_CLOUD_PROJECT")
+LOCATION = "us-central1"
+vertexai.init(project=PROJECT_ID, location=LOCATION)
+
+
+@app.route("/static")
+def get_static_coordinates():
+    """
+    Retrieves coordinates for a static address using the Vertex AI Generative AI API and the Nominatim API.
+
+    This function uses a pre-defined prompt to request coordinates for a specific address.
+    It then extracts the arguments from the function call response and constructs a URL for the Nominatim API.
+    Finally, it retrieves the coordinates from the Nominatim API and returns them as a JSON object.
+    """
+
+    prompt = """
+   I want to get the coordinates for the following address:
+   1600 Amphitheatre Pkwy, Mountain View, CA 94043, US
+   """
+
+    response = model.generate_content(
+        prompt,
+        generation_config=GenerationConfig(temperature=0),
+        tools=[location_tool],
+    )
+    response.candidates[0].content.parts[0]
+
+    x = response.candidates[0].content.parts[0].function_call.args
+
+    url = "https://nominatim.openstreetmap.org/search?"
+    for i in x:
+        url += '{}="{}"&'.format(i, x[i])
+    url += "format=json"
+
+    headers = {
+        "User-Agent": "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"
+    }
+    x = requests.get(url, headers=headers)
+    content = x.json()
+    return content
```

Comment:
Maybe this part could be wrapped up in a function and called with the static location, then that function could be reused for the `get_coordinates()` method.

Reply:
i think this part is only useful for the blog post, i can probably just remove it. it adds complexity to the code sample.